mmark.1
.\" Automatically generated by Pandoc 1.17.2
.\"
.TH "MMARK" "1" "August 2018" "" ""
.hy
.SH NAME
.PP
mmark \[en] generate XML, HTML or markdown from mmark markdown
documents.
.SH SYNOPSIS
.PP
\f[B]mmark\f[] [\f[B]OPTIONS\f[]] [\f[I]FILE...\f[]]
.SH DESCRIPTION
.PP
\f[B]Mmark\f[] is a powerful markdown processor written in Go, geared
towards writing IETF documents.
It is, however, \f[I]also\f[] suited for writing complete books and
other technical documentation.
.PP
Mmark provides an advanced markdown dialect that processes file(s) to
produce internet\-drafts in XML RFC
7991 (https://tools.ietf.org/html/rfc7991) format.
Mmark can produce xml2rfc (aforementioned RFC 7991), RFC 7749 (xml2rfc
version 2), HTML5 and markdown output.
.PP
The syntax is detailed at <https://mmark.nl/syntax>.
.PP
Compared to other markdown variants mmark adds:
.IP \[bu] 2
(Extended) Title blokto specify authors and IETF specific bits.
.IP \[bu] 2
Special sections for abstracts, prefaces or notes.
.IP \[bu] 2
Including other files with the option to specify line ranges, regular
expressions and/or prefixing each line with a custom string.
.IP \[bu] 2
Document divisions: front matter, main matter and back matter.
.IP \[bu] 2
Captions for code, tables, quotes and subfigures.
.IP \[bu] 2
Asides for small notes.
.IP \[bu] 2
Figures and Subfigures that allow for grouping images into subfigures as
well as giving a single image metadata (a link, attributes, etc.).
.IP \[bu] 2
Block Level Attributes that allow to specify attributes, classes and IDs
for elements.
.IP \[bu] 2
Indices to mark an item (and/or a subitem) to be referenced in the
document index.
.IP \[bu] 2
Citations and adding XML References (those used by the IETF).
.IP \[bu] 2
Cross references: short form of referencing a section in the document.
.IP \[bu] 2
Super\- and Subscript.
.IP \[bu] 2
Callouts in code and text.
.IP \[bu] 2
BCP14 (RFC 2119) keyword detection.
.SS RFC 7749
.PP
This is currently the XML format used by the RFC editor for accepting
Internet\-Drafts.
Some of these turn into RFCs.
For getting text output you\[aq]ll need xml2rfc for the actual
conversion.
.SS RFC 7991
.PP
This is the future XML format used by the RFC editor for accepting
Internet\-Drafts.
A valid markdown document can be turned in RFC 7749 or RFC 7991 XML.
.SS HTML5
.PP
The HTML5 renderer outputs HTML.
.SS Markdown
.PP
Mmark can also be "translated" into markdown again.
This is a useful feature for auto\-formatting markdown files.
.SH OPTIONS
.TP
.B \f[B]\-ast\f[]
print abstract syntax tree and exit.
.RS
.RE
.TP
.B \f[B]\-css string\f[]
link to a CSS stylesheet (only used with \-html).
.RS
.RE
.TP
.B \f[B]\-fragment\f[]
don\[aq]t create a full document.
.RS
.RE
.TP
.B \f[B]\-head string\f[]
link to HTML to be included in head (only used with \-html).
.RS
.RE
.TP
.B \f[B]\-html\f[]
create HTML output.
.RS
.RE
.TP
.B \f[B]\-2\f[]
generate RFC 7749 XML.
.RS
.RE
.TP
.B \f[B]\-markdown\f[]
output (normalized) markdown.
.RS
.RE
.TP
.B \f[B]\-text\f[]
output text with ANSI codes (aim to be somewhat similar to a manual
page).
.RS
.RE
.TP
.B \f[B]\-unsafe\f[]
allow includes from anywhere in the filesystem, otherwise they are only
allowed \f[I]under\f[] the current document.
.RS
.RE
.TP
.B \f[B]\-textwidth integer\f[]
set the text width when generating markdown, defaults to 100 characters.
.RS
.RE
.TP
.B \f[B]\-w\f[]
write to source file when generating markdown.
.RS
.RE
.TP
.B \f[B]\-index\f[]
generate an index at the end of the document (default true).
.RS
.RE
.TP
.B \f[B]\-bibliography\f[]
generate a bibliography section after the back matter (default true),
this needs a \f[C]{{backmatter}}\f[] in the document.
.RS
.RE
.TP
.B \f[B]\-version\f[]
show mmark\[aq]s version.
.RS
.RE
.SH ALSO SEE
.PP
RFC 7991 and RFC 7749.
The main site for Mmark is mmark.nl (https://mmark.nl).
.SH AUTHORS
Mmark Authors.


mmark.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"

	"github.com/gomarkdown/markdown"
	"github.com/gomarkdown/markdown/ast"
	"github.com/gomarkdown/markdown/html"
	"github.com/gomarkdown/markdown/parser"
	"github.com/mmarkdown/mmark/mast"
	"github.com/mmarkdown/mmark/mparser"
	mmarkout "github.com/mmarkdown/mmark/render/markdown"
	"github.com/mmarkdown/mmark/render/mhtml"
	"github.com/mmarkdown/mmark/render/text"
	"github.com/mmarkdown/mmark/render/xml"
	"github.com/mmarkdown/mmark/render/xml2"
)

var (
	flagCSS      = flag.String("css", "", "link to a CSS stylesheet (only used with -html)")
	flagHead     = flag.String("head", "", "link to HTML to be included in head (only used with -html)")
	flagAst      = flag.Bool("ast", false, "print abstract syntax tree and exit")
	flagBib      = flag.Bool("bibliography", true, "generate a bibliography section after the back matter")
	flagFragment = flag.Bool("fragment", false, "don't create a full document")
	flagHTML     = flag.Bool("html", false, "create HTML output")
	flagIndex    = flag.Bool("index", true, "generate an index at the end of the document")
	flagTwo      = flag.Bool("2", false, "generate RFC 7749 XML")
	flagMarkdown = flag.Bool("markdown", false, "generate markdown (experimental)")
	flagText     = flag.Bool("text", false, "generate text for ANSI escapes")
	flagWrite    = flag.Bool("w", false, "write to source file when generating markdown")
	flagWidth    = flag.Int("width", 100, "text width when generating markdown")
	flagUnsafe   = flag.Bool("unsafe", false, "allow unsafe includes")
	flagVersion  = flag.Bool("version", false, "show mmark version")
)

func main() {
	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "SYNOPSIS: %s [OPTIONS] %s\n", os.Args[0], "[FILE...]")
		fmt.Println("\nOPTIONS:")
		flag.PrintDefaults()
	}

	flag.Parse()
	args := flag.Args()
	if len(args) == 0 {
		args = []string{"os.Stdin"}
	}
	if *flagVersion {
		fmt.Println(Version)
		os.Exit(0)
	}

	for _, fileName := range args {
		var (
			d    []byte
			err  error
			init mparser.Initial
		)
		if fileName == "os.Stdin" {
			init = mparser.NewInitial("")
			d, err = ioutil.ReadAll(os.Stdin)
			if err != nil {
				log.Printf("Couldn't read %q: %q", fileName, err)
				continue
			}
		} else {
			init = mparser.NewInitial(fileName)
			d, err = ioutil.ReadFile(fileName)
			if err != nil {
				log.Printf("Couldn't open %q: %q", fileName, err)
				continue
			}
		}
		if *flagUnsafe {
			init.Flags |= mparser.UnsafeInclude
		}

		if *flagMarkdown {
			mparser.Extensions &^= parser.Includes
		}

		p := parser.NewWithExtensions(mparser.Extensions)
		parserFlags := parser.FlagsNone
		documentTitle := "" // hack to get document title from toml title block and then set it here.
		if !*flagHTML {
			parserFlags |= parser.SkipFootnoteList // both xml formats don't deal with footnotes well.
		}
		p.Opts = parser.ParserOptions{
			ParserHook: func(data []byte) (ast.Node, []byte, int) {
				node, data, consumed := mparser.Hook(data)
				if t, ok := node.(*mast.Title); ok {
					if !t.IsTriggerDash() {
						documentTitle = t.TitleData.Title
					}
				}
				return node, data, consumed
			},
			ReadIncludeFn: init.ReadInclude,
			Flags:         parserFlags,
		}

		doc := markdown.Parse(d, p)
		if *flagBib {
			mparser.AddBibliography(doc)
		}
		if *flagIndex {
			mparser.AddIndex(doc)
		}

		if *flagAst {
			ast.Print(os.Stdout, doc)
			fmt.Print("\n")
			return
		}

		var renderer markdown.Renderer

		switch {
		case *flagHTML:
			opts := html.RendererOptions{
				// TODO(miek): make this an option.
				Comments:       [][]byte{[]byte("//"), []byte("#")},
				RenderNodeHook: mhtml.RenderHook,
				Flags:          html.CommonFlags | html.FootnoteNoHRTag | html.FootnoteReturnLinks,
				Generator:      `  <meta name="GENERATOR" content="github.com/mmarkdown/mmark Mmark Markdown Processor - mmark.nl`,
			}
			if !*flagFragment {
				opts.Flags |= html.CompletePage
			}
			opts.CSS = *flagCSS
			if *flagHead != "" {
				head, err := ioutil.ReadFile(*flagHead)
				if err != nil {
					log.Printf("Couldn't open %q, error: %q", *flagHead, err)
					continue
				}
				opts.Head = head
			}
			if documentTitle != "" {
				opts.Title = documentTitle
			}

			renderer = html.NewRenderer(opts)
		case *flagTwo:
			opts := xml2.RendererOptions{
				Flags:    xml2.CommonFlags,
				Comments: [][]byte{[]byte("//"), []byte("#")},
			}
			if *flagFragment {
				opts.Flags |= xml2.XMLFragment
			}

			renderer = xml2.NewRenderer(opts)
		case *flagMarkdown:
			opts := mmarkout.RendererOptions{TextWidth: *flagWidth}
			renderer = mmarkout.NewRenderer(opts)
		case *flagText:
			opts := text.RendererOptions{TextWidth: *flagWidth}
			renderer = text.NewRenderer(opts)
		default:
			opts := xml.RendererOptions{
				Flags:    xml.CommonFlags,
				Comments: [][]byte{[]byte("//"), []byte("#")},
			}
			if *flagFragment {
				opts.Flags |= xml.XMLFragment
			}

			renderer = xml.NewRenderer(opts)
		}

		x := markdown.Render(doc, renderer)
		if *flagMarkdown && *flagWrite && fileName != "os.Stdin" {
			ioutil.WriteFile(fileName, x, 0600)
			continue
		}
		if *flagMarkdown {
			fmt.Print(string(x))
			continue
		}

		fmt.Println(string(x))
	}
}
